Deal with duplicate errors more gracefully

wendi-yu opened the pull request:
The error list is now an error set to prevent the user being inundated
by a ton of repeated instances caused by the same error. Instead, if
an identical duplicate is detected, a single "note" error is added to
the error set to indicate there were multiple occurrences of this error.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9)
<!-- Reviewable:end -->

Commit:
Deal with duplicate errors more gracefully

The error list is now an error set to prevent the user being inundated
by a ton of repeated instances caused by the same error. Instead, if
an identical duplicate is detected, a single "note" error is added to
the error set to indicate there were multiple occurrences of this error.

wendi-yu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9)* status: 0 of 6 files reviewed, 1 unresolved discussion (waiting on @jacobdeery)

___
*[operations_simulator/plumbing/plumbing_engine.py, line 151 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-M4l6bM-9ChVJoZmMdeO:-M4l6bM-9ChVJoZmMdeP:bqgewr2) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/f406cf8f4a8cdc97a90bd8edbc8d15a03c9fffc6/operations_simulator/plumbing/plumbing_engine.py#L151)):*
> ```Python
>                         utils.multi_error_msg(
>                             f"Component '{component.name}', node {node} not found in mapping dict."),
>                             component.name, node)
> ```

There's a hanging indentation here, I'll fix it in the next commit/before this gets merged.


<!-- Sent from Reviewable.io -->

jacobdeery reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9)* status: 0 of 6 files reviewed, 5 unresolved discussions (waiting on @jacobdeery and @wendi-yu)

___
*[operations_simulator/plumbing/plumbing_engine.py, line 49 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-M4l7yXACZ7ujuUlWkls:-M4l7yXACZ7ujuUlWklt:bgs79v0) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/f406cf8f4a8cdc97a90bd8edbc8d15a03c9fffc6/operations_simulator/plumbing/plumbing_engine.py#L49)):*
<details><summary><i alt="summary">Quoted 9 lines of code…</i></summary>

> ```
>                 error = invalid.InvalidComponentName(
>                     f"Component with name '{name}' not found in provided mapping dict.", name)
>                 if error in self.error_set:
>                     error = invalid.InvalidComponentName(
>                         utils.multi_error_msg(
>                             "Component with name '{name}' not found in provided mapping dict."),
>                         name)
> 
>                 self.error_set.add(error)
> ```
</details>


The logic for this piece of code is repeated three times; it would be nice if we could move it into a separate function. With the way it's written right now (creating a new error of the same type), I'm not sure of the best way to do that.
___
*[operations_simulator/plumbing/plumbing_engine.py, line 65 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-M4l7jG-8jNoAqhMUr-g:-M4l7jG-8jNoAqhMUr-h:b-dwcwiu) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/f406cf8f4a8cdc97a90bd8edbc8d15a03c9fffc6/operations_simulator/plumbing/plumbing_engine.py#L65)):*
> ```Python
>                     self.plumbing_graph.add_edge(
>                         nodes_map[start_node], nodes_map[end_node], edge_key)
>                 else:
> ```

Hmm, doesn't this make it so that if both `start_node` and `end_node` had errors, only one of those errors will be reported? Is this the desired behaviour?
___
*[operations_simulator/plumbing/plumbing_engine.py, line 143 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-M4lHnyGDXqeXacp3lol:-M4lHnyGDXqeXacp3lom:b-ya81ng) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/f406cf8f4a8cdc97a90bd8edbc8d15a03c9fffc6/operations_simulator/plumbing/plumbing_engine.py#L143)):*
> ```Python
>                 new_edge = (component_map[cstart_node], component_map[cend_node], key)
>                 state_edges_graph[new_edge] = state_edges_component[cedge]
>             else:
> ```

Same as above
___
*[operations_simulator/plumbing/plumbing_utils.py, line 38 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-M4l7TcC3sL4LNzR6Mxc:-M4l7TcC3sL4LNzR6Mxd:bibfqpg) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/f406cf8f4a8cdc97a90bd8edbc8d15a03c9fffc6/operations_simulator/plumbing/plumbing_utils.py#L38)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> def multi_error_msg(error):
>     message = "Multiple instances of this error were detected:\n"\
>         + error + "\n"\
>         + "Consider checking for errors in the mapping dict."
>     return message
> ```
</details>


I think maybe this function belongs in `invalid_reasons.py`


<!-- Sent from Reviewable.io -->

wendi-yu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9)* status: 0 of 6 files reviewed, 4 unresolved discussions (waiting on @jacobdeery)

___
*[operations_simulator/plumbing/plumbing_engine.py, line 49 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-M4l7yXACZ7ujuUlWkls:-M4ljemx39i4htzFTdzu:bt9m9s0) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/f406cf8f4a8cdc97a90bd8edbc8d15a03c9fffc6/operations_simulator/plumbing/plumbing_engine.py#L49)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
>                 error = invalid.InvalidComponentName(
>                     f"Component with name '{name}' not found in provided mapping dict.", name)
>                 if error in self.error_set:
>                     error = invalid.InvalidComponentName(
>                         utils.multi_error_msg(
>                             "Component with name '{name}' not found in provided mapping dict."),
>                         name)
> 
>                 self.error_set.add(error)
> ```

The logic for this piece of code is repeated three times; it would be nice if we could move it into a separate function. With the way it's written right now (creating a new error of the same type), I'm not sure of the best way to do that.
</blockquote></details>

Yep, implemented it with the new `DuplicateError` class
___
*[operations_simulator/plumbing/plumbing_engine.py, line 65 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-M4l7jG-8jNoAqhMUr-g:-M4lj6uE9HlNc-cqU9jh:byu413f) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/f406cf8f4a8cdc97a90bd8edbc8d15a03c9fffc6/operations_simulator/plumbing/plumbing_engine.py#L65)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

Hmm, doesn't this make it so that if both `start_node` and `end_node` had errors, only one of those errors will be reported? Is this the desired behaviour?
</blockquote></details>

Oh yep you're right, changed that back
___
*[operations_simulator/plumbing/plumbing_engine.py, line 143 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-M4lHnyGDXqeXacp3lol:-M4lj65431EjqpsbAI5L:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/f406cf8f4a8cdc97a90bd8edbc8d15a03c9fffc6/operations_simulator/plumbing/plumbing_engine.py#L143)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

Same as above
</blockquote></details>

Done.
___
*[operations_simulator/plumbing/plumbing_utils.py, line 38 at r1](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-M4l7TcC3sL4LNzR6Mxc:-M4ljoiVBWKAgGGoL7Eb:b-896fix) ([raw file](https://github.com/waterloo-rocketry/operations-simulator/blob/f406cf8f4a8cdc97a90bd8edbc8d15a03c9fffc6/operations_simulator/plumbing/plumbing_utils.py#L38)):*
<details><summary><i>Previously, jacobdeery (Jacob Deery) wrote…</i></summary><blockquote>

> ```
> def multi_error_msg(error):
>     message = "Multiple instances of this error were detected:\n"\
>         + error + "\n"\
>         + "Consider checking for errors in the mapping dict."
>     return message
> ```

I think maybe this function belongs in `invalid_reasons.py`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
Implement specific duplicate errors class

Duplicate errors are now stored in their own class, rather than being
created as another instance of the class of error being duplicated.
Also some code cleanup.

jacobdeery reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/operations-simulator/9#-:-M4qq1ryE3FxIr4FoBL2:bnfp4nl)* status: 0 of 5 files reviewed, 4 unresolved discussions (waiting on @jacobdeery)



<!-- Sent from Reviewable.io -->
